ingestion-edge/tests/flush_manager/test_flush_manager.py

```diff
@@ -0,0 +1,174 @@
+
+
+def delete_pvcs(api: CoreV1Api):
+    for pvc in list_pvcs(api):
+        api.delete_namespaced_persistent_volume_claim(
+            pvc.metadata.name,
+            pvc.metadata.namespace,
+            body=V1DeleteOptions(
+                grace_period_seconds=0,
+                propagation_policy="Background",
+                preconditions=V1Preconditions(
+                    resource_version=pvc.metadata.resource_version,
+                    uid=pvc.metadata.uid,
+                ),
+            ),
+        )
+
+
+def restart_web_pods(api: CoreV1Api):
+    for pod in api.list_namespaced_pod("default").items:
+        if pod.metadata.labels.get("app") != "web":
+            continue
+        api.delete_namespaced_pod(
+            pod.metadata.name,
+            pod.metadata.namespace,
+            body=V1DeleteOptions(
+                grace_period_seconds=0,
+                propagation_policy="Background",
+                preconditions=V1Preconditions(
+                    resource_version=pod.metadata.resource_version,
+                    uid=pod.metadata.uid,
+                ),
+            ),
+        )
+
+
+def test_flush_manager(options: Dict[str, Any], emulator: str, web: str):
+    print("starting test")
+    api = CoreV1Api()
+    # max number of loops to run when waiting for kube actions to complete
+    max_wait_loops = 20 if options["cluster"] is None else 60
+
+    # server has invalid PUBSUB_EMULATOR, so that only flush can deliver messages
+    static_pvs = options["cluster"] is None
+    if static_pvs:
+        create_static_pvs(api)
+
+    print("waiting for pods to be healthy")
+    for _ in range(max_wait_loops):
+        if all(
+            pod.status.phase == "Running"
+            for pod in api.list_namespaced_pod("default").items
+        ):
+            break
+        time.sleep(1)
+    else:
+        assert False, "pods did not become healthy"
+
+    # create a subscription to the defined topic
+    print("creating pubsub subscription")
+    os.environ["PUBSUB_EMULATOR_HOST"] = emulator
+    sub_client = SubscriberClient()
+    topic_path = "projects/{project}/topics/{topic}".format(**options)
+    subscription_path = "projects/{project}/subscriptions/{topic}".format(**options)
+    try:
+        sub_client.create_subscription(subscription_path, topic_path, retry=None)
+    except AlreadyExists:
+        pass
+
+    print("posting message 0")
+    requests.post(web, headers={"host": "web"}, json={"id": 0}).raise_for_status()
+    print("setting up race condition: attached pvc is also deleted")
+    delete_pvcs(api)
+    print("setting up race condition: pod unschedulable due to missing pvc")
+    with pytest.raises(ApiException) as excinfo:
+        restart_web_pods(api)
+    assert excinfo.value.reason == "Conflict"
+    print("posting message 1")
+    with pytest.raises(requests.exceptions.ConnectionError):
+        requests.post(web, headers={"host": "web"}, json={"id": 1}).raise_for_status()
+
```

Comment:
If the flush manager were not run and the pod were recreated, what is the behavior that's undesirable? Resending of a message that has already been flushed once?

Reply:
if you are asking about prod: messages that are queued to disk and not delivered before shutdown will sit on unattached disks until the stateful set is scaled back up, at which time it will re-attach the disk and flush the messages. this can take days, resulting in very-late delivery of messages.

if you asking about this test: the web pods are configured with an invalid `PUBSUB_EMULATOR_HOST` so that messages are always queued, and only delivered by the jobs the flush manager creates. if the flush manager is not run then the messages are not flushed at all, and the only undesirable behavior I can think of that you might be asking about is that the test would not cover the full behavior of the flush manager.